docs/okeps/okep-5085-localnet-api.md

```diff
@@ -0,0 +1,834 @@
+additional developments for improving how this state is communicated to users. 
+Please see the [Alternatives](#indicating-pod-configured-with-outdated-network-config) section for more details.
+
+#### Workflow Description
+
+The CUDN CRD controller should be changed accordingly to support localnet topology.
+It should validate localnet topology configuration and generate corresponding NADs for localnet as other topologies (Layer2 & Layer3)
+in the selected namespaces.
+
+As of today the CUDN CRD controller does not allow mutating `spec.network`.
+
+OVN-Kubernetes support NAD spec mutations on day 2.
+For the new network configuration (defined in the spec) to take place, the user must restart the workloads.
+
+In order to comply with NAD spec mutation support and provide better UX,
+the CUDN CRD controller should allow `spec.network` mutations for localnet topologies.
+Enable changing at least the following fields: `MTU`, `VLAN`, `excludeSubnets` and `physicalNetworkName`.
+
+In any order:
+- The user configures localnet bridge mapping on the nodes, e.g.: using NNCP.
+- Create CUDN, specifying the bridge-mapping network name in the spec.
+
+Allowing users to mutate the mentioned settings on day2 enables them to act on environment changes that require OVN network setting change,
+saving the hassle of re-creating the whole project (delete pods, delete network CR, create the network CR with new settings, create pods ...).
+
+Naturally, changing network setting while workloads connected to it has implications, may introduce transient network disruption, 
+which manifests in packet-drops, latency and disconnections.
+
+Having said that mutating the suggested settings should not break the workloads networking, and they should be able to communicate
+with each other.
+After the network setting changes, restarting the workloads is needed in order for them to consume the latest network-config.
+Once all connected workloads configured with the latest network-config the state should be reconciled and no network-disruption is expected.
+
+Please refer to [Risks Known Limitations and Mitigations](#risks-known-limitations-and-mitigations) section for more details.  
+
+When a CUDN CR localnet spec changes, and the change require updating the corresponding NAD spec.config, the controller should 
+add the proposed annotation `k8s.ovn.io/network-out-of-sync` to all connected pods in the selected-namespaces.
+In case the annotation is already exists and has items (networks), the controller should append the subject network to the list.
+
+On NAD update the controller should record the update timestamp, for each connected pod and compare the pod creation-timestamp and the NAD update timestamp:
+If the pod was created before NAD was updated - add the annotation to the pod.
+When pod created after NAD update, no need to annotate the pod because it consumed the latest network-config.
```

Comment:
Continuing the previous [thread](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/5087#discussion_r2004993490), this is still raceful.

Even if we assume the time known to the UDN controller is in total sync with the on of the api-server, I cannot see how you can assure no pod will enter in-between the steps of NAD update, timestamp saving and processing the pod list. The whole system is async based, nothing is atomic and all actual states are in constant potential transition. The only assured thing that exist is that at some future, we may be able to reach a desired state, and I do not see how this mechanism allows something like this.

Reply:
We decided to drop spec mutations and buy that the need to indicate users upon network-config change.
The proposal now focus on CRD extensions for localnet topology only.